plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptions.php

```diff
@@ -30,6 +30,13 @@ class AddToCartWithOptions extends AbstractBlock {
 	 */
 	protected function initialize() {
 		parent::initialize();
+
+		// Register a core/button variation for the Add To Cart Button block.
+		add_filter( 'register_block_type_args', array( $this, 'add_core_button_variation_for_add_to_cart_button' ), 10, 2 );
+
+		// Extend core/block rendering.
+		add_filter( 'render_block', array( $this, 'render_product_add_to_cart_with_options_button' ), 10, 3 );
```

Comment:
I have mixed feelings about this. To me, hooking into `render_block` is dangerous and I worry it might break in future updates. I wonder if that's a signal that this should be a completely different block (or we should look for an alternative approach).

I know the goal is to leverage the upstream block (and the associated styles), but I'm not convinced that hooking into `render_block` is a better approach than what we are doing in other places about using the CSS class name in a different block.

What are your thoughts on this?

Reply:
> To me, hooking into render_block is dangerous, and I worry it might break in future updates.

It's a fair point. However, being aware of the impact of extending, we'll be okay.
Extending what the API supports the block renders 🤷‍♂️ .
Also, it only introduces a few changes when rendering the block.
Even more, it enqueues the frontend script, something that we should move to another place definitely.

What's relevant is how it populates the markup with the interactivity attributes. This part is safe if I'm right.

> I wonder if that's a signal that this should be a completely different block (or we should look for an alternative approach).

There are a few challenges on the horizon that make me doubt:

* Pick the post (product) ID from the block context. The quick solution is extending the block type populating the `usesContext` array
* Hide the `Link` toolbar button. This can be done by binding the `URL` block attribute to an external source. Currently, the URL block attribute is the core post-meta source. However, it could also deal with product types like External Products.

> What are your thoughts on this?

I'd give it a chance, but remember that we'll probably have to use/create a custom block.

Reply:
This is the most relevant part so far.

```php
$processor = new \WP_HTML_Tag_Processor( $block_content );

if ( $processor->next_tag(
  array(
    'tag_name'   => 'div',
    'class_name' => 'wc-block-product-add-to-cart-with-options-button',
  )
) ) {
  $processor->set_attribute( 'data-wc-interactive', $data_wc_interactive );

  if ( $processor->next_tag( array( 'tag_name' => 'a', 'class_name' => 'wp-element-button' ) ) ) {
    $processor->set_attribute( 'data-wc-on--click', $data_wc_add_to_cart_action );
  }
}

return $processor->get_updated_html();
```